File: app/src/main/java/com/jdelorenzo/congressapp/ui/legislators/LegislatorViewModel.java
package com.jdelorenzo.congressapp.ui.legislators;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.LiveData;

import com.jdelorenzo.congressapp.data.LegislatorRepository;
import com.jdelorenzo.congressapp.data.model.Legislator;

import java.util.List;

/**
 *  {@link AndroidViewModel} for displaying a {@link Legislator}
 */
public class LegislatorViewModel extends AndroidViewModel {
    private LegislatorRepository legislatorRepository;
    private LiveData<List<Legislator>> legislators;

    public LegislatorViewModel(Application application) {
        super(application);
        legislatorRepository = new LegislatorRepository(application);
        legislators = legislatorRepository.getLegislators();
    }

    public LiveData<List<Legislator>> getLegislators() { return legislators; }
}

File: app/src/main/java/com/jdelorenzo/congressapp/ui/legislators/LegislatorViewModel.kt
package com.jdelorenzo.congressapp.ui.legislators

import android.arch.lifecycle.ViewModel

import com.jdelorenzo.congressapp.data.LegislatorDataSource
import com.jdelorenzo.congressapp.data.model.Legislator

import io.reactivex.Flowable

class LegislatorViewModel(private val dataSource: LegislatorDataSource) : ViewModel() {

    private var legislator: Legislator? = null

    /**
     * Get the name of a legislator
     *
     * @return a [io.reactivex.Flowable] that will emit every time the legislator's name
     * is updated.
     */
    val legislatorName: Flowable<String>
        get() = dataSource.legislator
                .map { legislator ->
                    this.legislator = legislator
                    legislator.firstName
                }
}
